filament/src/ShadowMapManager.cpp

```diff
@@ -193,47 +71,166 @@ void ShadowMapManager::addSpotShadowMap(size_t lightIndex) noexcept {
+
+    const bool fillWithCheckerboard = engine.debug.shadowmap.checkerboard && !view.hasVsm();
+
+    auto& shadowPass = fg.addPass<ShadowPassData>("Shadow Pass",
+            [&](FrameGraph::Builder& builder, auto& data) {
+                FrameGraphTexture::Descriptor shadowTexture {
+                    .width = mTextureRequirements.size, .height = mTextureRequirements.size,
+                    .depth = mTextureRequirements.layers,
+                    .levels = 1,
+                    .type = SamplerType::SAMPLER_2D_ARRAY,
+                    .format = mTextureFormat,
+                    .usage = TextureUsage::DEPTH_ATTACHMENT | TextureUsage::SAMPLEABLE
+                        | (fillWithCheckerboard ? TextureUsage::UPLOADABLE : (TextureUsage) 0)
+                };
+
+                if (view.hasVsm()) {
+                    // TODO: support 16-bit VSM depth textures.
+                    shadowTexture.format = TextureFormat::RG32F;
+                    shadowTexture.usage = TextureUsage::COLOR_ATTACHMENT | TextureUsage::SAMPLEABLE;
+                }
+
+                data.shadow = builder.createTexture("Shadow Texture", shadowTexture);
+                data.shadow = builder.write(data.shadow);
+
+                if (view.hasVsm()) {
+                    data.tempShadow = builder.createTexture("Temporary Shadow Texture", {
+                        .width = mTextureRequirements.size, .height = mTextureRequirements.size,
+                        .depth = 1,
+                        .levels = 1,
+                        .type = SamplerType::SAMPLER_2D,
+                        .format = TextureFormat::DEPTH16,
+                        .usage = TextureUsage::DEPTH_ATTACHMENT
+                    });
+                    // We specify "read" for the temporary shadow texture, so it isn't culled.
+                    data.tempShadow = builder.write(builder.read(data.tempShadow));
+                }
+
+                // Create a render target for each layer of the texture array.
+                for (uint8_t i = 0u; i < mTextureRequirements.layers; i++) {
+                    FrameGraphRenderTarget::Descriptor renderTarget {};
+                    if (view.hasVsm()) {
+                        renderTarget.attachments = { { data.shadow, 0u, i }, { data.tempShadow } };
+                        renderTarget.clearFlags = TargetBufferFlags::COLOR | TargetBufferFlags::DEPTH;
+                    } else {
+                        renderTarget.attachments = { {}, { data.shadow, 0u, i } };
+                        renderTarget.clearFlags = TargetBufferFlags::DEPTH;
+                    }
+                    data.rt[i] = builder.createRenderTarget("Shadow RT", renderTarget);
+                }
+            },
+            [=, passes = std::move(passes), &view](FrameGraphPassResources const& resources,
+                    auto const& data, DriverApi& driver) mutable {
+                if (UTILS_UNLIKELY(fillWithCheckerboard)) {
```

Comment:
you shouldn't have to do this, because this pass should be culled automatically. Is it not the case?

Reply:
Let me check. I guess it should be culled because I've only specified "write" in the "Shadow Debug Pattern Pass".